Initial version of class profiling for PGO

AndyAyersMS opened the pull request:
Add support to the jit and runtime so that PGO can determine the distribution of
classes at virtual and indirect call sites.

Use this information when jitting to enable guarded devirtualization, if there
is a suitably likely class to guess for.

Enable by setting:
```
COMPlus_TieredCompilation=1
COMPlus_TieredPGO=1
COMPlus_JitClassProfiling=1
COMPlus_JitEnableGuardedDevirtualization=1
```
impact can be enhanced by also setting
```
COMPlus_TC_QuickJitForLoops=1
```
to allow more methods to pass through Tier0.

AndyAyersMS committed:
Initial version of class profiling for PGO

Add support to the jit and runtime so that PGO can determine the distribution of
classes at virtual and indirect call sites.

Use this information when jitting to enable guarded devirtualization, if there
is a suitably likely class to guess for.

Enable by setting:
```
COMPlus_TieredCompilation=1
COMPlus_TieredPGO=1
COMPlus_JitClassProfiling=1
COMPlus_JitEnableGuardedDevirtualization=1
```
impact can be enhanced by also setting
```
COMPlus_TC_QuickJitForLoops=1
```
to allow more methods to pass through Tier0.

AndyAyersMS commented:
cc @dotnet/jit-contrib @davidwrighton 

Jit and jit interface changes should be fairly close to their final form. Runtime side will eventually be replaced as we build out more of the PGO story over there. But this is a start.

Still missing: crossgen2 "support", and some way to invalidate class profile entries for unloadable classes.

EgorBo commented on src/coreclr/src/ToolBox/superpmi/superpmi-shared/methodcontext.h:
```diff
@@ -1458,6 +1477,7 @@ enum mcPackets
     Packet_GetJitFlags                                   = 154, // Added 2/3/2016
     Packet_GetJitTimeLogFilename                         = 67,
     Packet_GetJustMyCodeHandle                           = 68,
+    Packet_GetLikelyClass                                = 181, // Added 9/27/2020
```
I guess it should be `182` (and comment `// Highest packet number: 181` should be updated)

AndyAyersMS replied:
Thanks. Seems like this should have caused a merge conflict.

AndyAyersMS committed:
fix SPMI packet numbers

BruceForstall commented on src/coreclr/src/vm/pgo.cpp:
```diff
@@ -342,6 +483,166 @@ HRESULT PgoManager::getMethodBlockCounts(MethodDesc* pMD, unsigned ilSize, UINT3
     return E_NOTIMPL;
 }
 
+CORINFO_CLASS_HANDLE PgoManager::getLikelyClass(MethodDesc* pMD, unsigned ilSize, unsigned ilOffset, UINT32* pLikelihood, UINT32* pNumberOfClasses)
```
Function header comment?

AndyAyersMS replied:
I'll add one, but they're not as common in runtime code as they are in the jit.

BruceForstall commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -11924,6 +11924,54 @@ HRESULT CEEJitInfo::getMethodBlockCounts (
+
+#ifdef FEATURE_PGO
+
+    // If TieredPGO is enabled, query the per call site class profile.
```
The comment says "If TieredPGO is enabled," but there is no check for TieredPGO in this function.

AndyAyersMS replied:
I'll reword that; if TieredPGO is not enabled the pgo manager won't find any class profiles.

BruceForstall commented on src/coreclr/src/jit/block.h:
```diff
@@ -448,6 +448,7 @@ struct BasicBlock : private LIR::Range
 
 #define BBF_BACKWARD_JUMP_TARGET           MAKE_BBFLAG(36) // Block is a target of a backward jump
 #define BBF_PATCHPOINT                     MAKE_BBFLAG(37) // Block is a patchpoint
+#define BBF_HAS_VIRTUAL_CALL               MAKE_BBFLAG(38) // BB contains a call needing a class profile
```
Aren't you doing profiling on both virtual and interface calls? Should this be named `BBF_NEEDS_CLASS_PROFILE`?

BruceForstall commented on src/coreclr/src/jit/flowgraph.cpp:
```diff
@@ -398,45 +420,254 @@ void Compiler::fgInstrumentMethod()
         countOfBlocks++;
     }
 
+    // We've alread counted the number of class probes
```
typo: alread

BruceForstall commented on src/coreclr/src/jit/flowgraph.cpp:
```diff
@@ -398,45 +420,254 @@ void Compiler::fgInstrumentMethod()
+    //
+    int countOfCalls = info.compClassProbeCount;
+
+    // Optionally bail out, if there are less than three blocks an no call sites to profile
```
grammar: an => and (and add trailing period)

BruceForstall commented on src/coreclr/src/jit/flowgraph.cpp:
```diff
@@ -398,45 +420,254 @@ void Compiler::fgInstrumentMethod()
+    int countOfCalls = info.compClassProbeCount;
+
+    // Optionally bail out, if there are less than three blocks an no call sites to profile
+    // One block is common. We don't expect to see zero or two blcoks here.
```
typo: blcoks

AndyAyersMS commented:
Not surprisingly, crossgen2 is unhappy; I need to stub out `getLikelyClass`.

BruceForstall commented on src/coreclr/src/jit/flowgraph.cpp:
```diff
@@ -398,45 +420,254 @@ void Compiler::fgInstrumentMethod()
+        }
 
-        for (block = fgFirstBB; (block != nullptr); block = block->bbNext)
+        // We may see class probes in internal blcoks, thanks to the
```
typo: blcoks

AndyAyersMS committed:
stub out crossgen2 support

jkotas commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5233,6 +5234,73 @@ void JIT_Patchpoint(int* counter, int ilOffset)
+    VALIDATEOBJECTREF(objRef);
+
+    ICorJitInfo::ClassProfile* const classProfile = (ICorJitInfo::ClassProfile*) tableAddress;
+    const int count = classProfile->Count++;
```
Does this need to use volatile and/or interlocked operations?

In `!_DEBUG`, it is a valid C/C++ optimization to get rid of the `count` local and re-fetch it below, ie transform the line below to `classProfile->ClassTable[classProfile->Count]`. I would potentially lead to buffer overruns or skipped slots in `ClassProfile`

AndyAyersMS replied:
Yes it should be volatile so it's just read once. Likewise for `s_rng`.

It is OK to drop updates because of races, and we're trying to keep overhead to a minimum, so no need to interlock.

BruceForstall reviewed (approved):
A few comments/questions

AndyAyersMS commented on src/coreclr/src/inc/corjit.h:
```diff
@@ -251,6 +251,14 @@ class ICorJitInfo : public ICorDynamicInfo
+    struct ClassProfile
+    {
+        enum { SIZE = 8, SAMPLE_INTERVAL = 32 };
+        UINT32 ILOffset;
```
Sure, will do. It's currently done so that we can allocate class profiles and block counts as one big chunk without telling the runtime how many of either exist. The high bit tells us this is a class profile, and the next bit which kind.

This magic will go away once we get more of the runtime side implemented and can specify a more flexible "schema" for the profile data.

AndyAyersMS commented on src/coreclr/src/inc/corjit.h:
```diff
@@ -251,6 +251,14 @@ class ICorJitInfo : public ICorDynamicInfo
 
+    struct ClassProfile
+    {
+        enum { SIZE = 8, SAMPLE_INTERVAL = 32 };
```
I'll add some description.

For count overflow I currently don't have great answers; it should be rare (we'd need methods that spent a lot of cycles in Tier0) but obviously it's not impossible. Right now overflow won't cause any crashes but it might lead to nonsensical profile data.

AndyAyersMS commented on src/coreclr/src/ToolBox/superpmi/superpmi-shared/icorjitinfoimpl.h:
```diff
@@ -993,6 +993,15 @@ HRESULT getMethodBlockCounts(CORINFO_METHOD_HANDLE ftnHnd,
                              BlockCounts**     pBlockCounts,
                              UINT32 *          pNumRuns);
 
+// Get the likely implementing class for a virtual call or interface call made by ftnHnd
```
Commented out, like BlockCounts is ?

BruceForstall replied:
> Commented out, like BlockCounts is ?

If necessary. I can't remember why BlockCounts is commented out; if it's required or not.

BruceForstall replied:
I guess if https://github.com/dotnet/runtime/pull/45044 is merged, it will be unnecessary.

AndyAyersMS committed:
review feedback

AndyAyersMS committed:
fix compile issue

AndyAyersMS committed:
crossgen2 getLikelyClass must return value

AndyAyersMS committed:
null, not 0

AndyAyersMS committed:
Restore the logic to do guarded devirt for predicatble interface
calls. Not sure how this got lost.

briansull commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5233,6 +5234,74 @@ void JIT_Patchpoint(int* counter, int ilOffset)
+    const unsigned count = *pCount++;
+    const unsigned S = ICorJitInfo::ClassProfile::SIZE;
+    const unsigned N = ICorJitInfo::ClassProfile::SAMPLE_INTERVAL;
+
```
assert(N >= S);  // SAMPLE_INTERVAL must be >= SIZE.

and possibly assert that N is a power of two

AndyAyersMS replied:
N doesn't need to a power of 2, though it probably will be. in practice.

briansull commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5233,6 +5234,74 @@ void JIT_Patchpoint(int* counter, int ilOffset)
+        //
+        if ((x % N) < S)
+        {
+            unsigned i = x % S;
```
For a divide by an unsigned constant the compiler would use a reciprocal multiply.  So this could be written using modulus and we wouldn't require S to be a power of two.  I think that the replacement of values in the ClassTable[] would also be more random when N is not divisible by S.

AndyAyersMS replied:
It is already written using a modulus...? 

In the most general form we should actually be checking `(x % count) < S` so the divisor would be variable. I'm experimenting with using a fixed window size here to bias the sample somewhat towards more recent observations.

Using a prime `S` (say) is an interesting idea, but `S` needs to be very small, so perhaps 7 or 11 might be worth a try.

briansull commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5233,6 +5234,74 @@ void JIT_Patchpoint(int* counter, int ilOffset)
 
 #endif // FEATURE_ON_STACK_REPLACEMENT
 
+HCIMPL2(void, JIT_ClassProfile, Object *obj, void* tableAddress)
```
Add an method header comment staying that this is a profiling/instrumentation method and needs to be optimized  to minimize the cost of runtime profiling.

We might want to explicitly tell the compiler to optimize this method (in retail builds) as we might not normally gather any C++ PGO data for this case if this feature it isn't enabled by default.

briansull commented on src/coreclr/src/vm/jithelpers.cpp:
```diff
@@ -5233,6 +5234,74 @@ void JIT_Patchpoint(int* counter, int ilOffset)
+        // intentionally simple so we can have multithreaded
+        // access w/o tearing state.
+        //
+        static volatile unsigned s_rng = 100;
```
If C++ supports thread local statics this would be a good use case.
Alternatively the runtime has support for per thread data, I believe.
Otherwise each thread has to grab the cache line for writing.

AndyAyersMS replied:
The class profile data is shared so I'm not sure there's much to be gained by having per-thread RNG state.

briansull commented on src/coreclr/src/vm/pgo.cpp:
```diff
@@ -8,11 +8,74 @@
+struct HistogramEntry
+{
+    CORINFO_CLASS_HANDLE m_mt;
+    unsigned             m_count;
```
Document the meanings of these fields:

m_count is the number of occurrences of m_mt in ClassTable[]

briansull commented on src/coreclr/src/vm/pgo.cpp:
```diff
@@ -342,6 +479,166 @@ HRESULT PgoManager::getMethodBlockCounts(MethodDesc* pMD, unsigned ilSize, UINT3
+            // Yep, found data. See if there is a suitable class profile.
+            //
+            // This bit is currently somewhat hacky ... we scan the records, the count records come
+            // first and are in increasing IL offset order. Class profiles have inverted IL offsets
```
What exactly is an "inverted" IL offset?

AndyAyersMS replied:
I set some flags bits in the high part of the IL offset so the runtime side can distinguish class profiles from regular block counts. 

This is a temporary measure needed until the runtime side evolves to understand there are different kinds of profile data.

briansull commented on src/coreclr/src/vm/pgo.cpp:
```diff
@@ -342,6 +479,166 @@ HRESULT PgoManager::getMethodBlockCounts(MethodDesc* pMD, unsigned ilSize, UINT3
+                    // Need to make sure this is even divisor
+                    //
+                    j += sizeof(ICorJitInfo::ClassProfile) / sizeof(ICorJitInfo::BlockCounts);
+                    continue;
```
I don't understand why we need to make sure this is an even divisor?

AndyAyersMS replied:
Because the underlying data is allocated in BlockCount increments, we need to make sure ClassProfile is a multiple, so that this update to `j` gets us to the next ClassProfile record.

briansull commented on src/coreclr/src/vm/pgo.cpp:
```diff
@@ -342,6 +479,166 @@ HRESULT PgoManager::getMethodBlockCounts(MethodDesc* pMD, unsigned ilSize, UINT3
 
+CORINFO_CLASS_HANDLE PgoManager::getLikelyClass(MethodDesc* pMD, unsigned ilSize, unsigned ilOffset, UINT32* pLikelihood, UINT32* pNumberOfClasses)
+{
+    *pLikelihood = 0;
```
pLikelyhood couls also be a 32-bit float.
As it currently stands is is an integer percentage, where 100 means 100% and 33 means 33%

AndyAyersMS replied:
I don't think we need that sort of precision; our histograms are approximate.

briansull commented on src/coreclr/src/vm/jitinterface.cpp:
```diff
@@ -11924,6 +11924,54 @@ HRESULT CEEJitInfo::getMethodBlockCounts (
+                     CORINFO_CLASS_HANDLE  baseHnd,
+                     UINT32                ilOffset,
+                     UINT32 *              pLikelihood,
+                     UINT32 *              pNumberOfClasses
```
pLikelyhood couls also be a 32-bit float.
As it currently stands it is an integer percentage, where 100 means 100% and 33 means 33%

AndyAyersMS replied:
Same goes here, histogram data is approximate.

What mostly matters is that for monomorphic types we get it right and predict high likelihood. For other cases the payoff is smaller.

AndyAyersMS commented on src/coreclr/src/inc/corjit.h:
```diff
@@ -251,6 +251,34 @@ class ICorJitInfo : public ICorDynamicInfo
+    {
+        enum { 
+            SIZE = 8, 
+            SAMPLE_INTERVAL = 32, 
```
They can be any positive value provided N >= S.

AndyAyersMS committed:
Generalize handling of the case where the dispatch object has a class type
to work with shared classes (like __Canon). Also note we can reach this
stage for both virtual and interface calls.

briansull reviewed (approved):
Looks Good

AndyAyersMS committed:
review feedback

AndyAyersMS commented:
I was looking at #4489 and realized it was a good test case for GDV via TieredPGO. Drilling in, I realized the logic I had for determining when to guess for a class wasn't as general as it could be and was missing out on guessing for calls dispatched through some shared classes (notably, `__Canon`).

So I've updated that part of `impDevirtualizeCall` and fixed a couple other small issues.

```
impDevirtualizeCall: Trying to devirtualize interface call:
    class for 'this' is __Canon (attrib 20020000)
    base method is ICalls::Execute
--- base class is interface
--- no derived method
    Class not final or exact
Consdering guarded devirt...
Likely class for 00007FF8C4640D10 (__Canon) is 00007FF8C48393E8 (ClassCalls) [likelihood:100 classes seen:1]
interface call would invoke method Execute
Marking call [000008] as guarded devirtualization candidate; will guess for class ClassCalls

...

Inlines into 06000003 [via DefaultPolicy] Runtime4489:UseInterfaceCalls():this
  [1 IL=0012 TR=000004 0600000E] [aggressive inline attribute] Executer`1:Execute(__Canon):this
    [2 IL=0008 TR=000024 0600000A] [aggressive inline attribute guarded devirt] ClassCalls:Execute():this
    [0 IL=0008 TR=000008 06000009] [FAILED: not inline candidate guarded] ICalls:Execute():this
```
resulting assembly code is:
```asm
; Assembly listing for method Runtime4489:UseInterfaceCalls():this
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; Tier-1 compilation
; optimized code
; optimized using profile data
; rsp based frame
; partially interruptible
; with IBC profile data, edge weights are valid, and fgCalledCount is 15837975
; invoked as altjit
; Final local variable assignments
;
;  V00 this         [V00,T00] (  4,  4   )     ref  ->  rcx         this class-hnd
;  V01 OutArgs      [V01    ] (  1,  1   )  lclBlk (32) [rsp+0x00]   "OutgoingArgSpace"
;  V02 tmp1         [V02,T02] (  2,  4   )     ref  ->  rcx         ld-addr-op class-hnd "Inlining Arg"
;  V03 tmp2         [V03,T01] (  3,  4   )     ref  ->  rcx         "guarded devirt this temp"
;* V04 tmp3         [V04    ] (  0,  0   )     ref  ->  zero-ref    class-hnd exact "guarded devirt this exact temp"
;  V05 tmp4         [V05,T03] (  2,  4   )     ref  ->  r11         class-hnd "Inlining Arg"
;
; Lcl frame size = 40

G_M18455_IG01:              ;; offset=0000H
       4883EC28             sub      rsp, 40
						;; bbWeight=1    PerfScore 0.25
G_M18455_IG02:              ;; offset=0004H
       4C8B5928             mov      r11, gword ptr [rcx+40]
       488B4910             mov      rcx, gword ptr [rcx+16]
       45391B               cmp      dword ptr [r11], r11d
       49BBF0684EC0F87F0000 mov      r11, 0x7FF8C04E68F0
       4C3919               cmp      qword ptr [rcx], r11       // GDV check
       7517                 jne      SHORT G_M18455_IG04
       48B918B54DC0F87F0000 mov      rcx, 0x7FF8C04DB518        // correct guess
       4533DB               xor      r11d, r11d
       448919               mov      dword ptr [rcx], r11d
       FF01                 inc      dword ptr [rcx]
						;; bbWeight=1    PerfScore 15.75
G_M18455_IG03:              ;; offset=0030H
       4883C428             add      rsp, 40
       C3                   ret      
						;; bbWeight=1    PerfScore 1.25
G_M18455_IG04:              ;; offset=0035H
       49BB580521C0F87F0000 mov      r11, 0x7FF8C0210558     // incorrect guess
       48B8580521C0F87F0000 mov      rax, 0x7FF8C0210558
       FF10                 call     qword ptr [rax]ICalls:Execute():this
       EBE3                 jmp      SHORT G_M18455_IG03
						;; bbWeight=0    PerfScore 0.00

```
I also verified I can drop a checked clrjit.dll next to the release one (calling it say clrjit.chk.dll) and invoke it selectively via the altjit mechanism; very handy for experimenting without having to swap jits around (and, drive all this via BDN).

kunalspathak commented:
> I also verified I can drop a checked clrjit.dll next to the release one (calling it say clrjit.chk.dll) and invoke it selectively via the altjit mechanism; very handy for experimenting without having to swap jits around (and, drive all this via BDN).

@DrewScoggins , @adamsitnik  - As we discussed this in the past, Andy confirmed that it works, so something we can try to get the disassembly more reliably?

AndyAyersMS commented:
Failure is known #45168. Mono build is a timeout (unrelated).
